[DDP] Support optim in backward after DDP init

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105995
* #105991

This allows in backward optimizers to be configured after DDP init, in
addition to before as was previously supported.

Differential Revision: [D47783347](https://our.internmc.facebook.com/intern/diff/D47783347/)

rohan-varma committed:
[DDP] Support optim in backward after DDP init

This allows in backward optimizers to be configured after DDP init, in
addition to before as was previously supported.

Differential Revision: [D47783347](https://our.internmc.facebook.com/intern/diff/D47783347/)

[ghstack-poisoned]

fegin reviewed (approved):
Do we expect any issue of lazy initialization? I guess not since `replicate` also uses lazy initialization. But just want to confirm.

rohan-varma commented:
@fegin I don't expect any issues and added some tests to verify the numerical parity.

rohan-varma committed:
Update on "[DDP] Support optim in backward after DDP init"


This allows in backward optimizers to be configured after DDP init, in
addition to before as was previously supported.

Differential Revision: [D47783347](https://our.internmc.facebook.com/intern/diff/D47783347/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support optim in backward after DDP init"


This allows in backward optimizers to be configured after DDP init, in
addition to before as was previously supported.

Differential Revision: [D47783347](https://our.internmc.facebook.com/intern/diff/D47783347/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
